### PR TITLE
Enable `on_before_batch_transfer` for `DPStrategy` and `IPUAccelerator`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Call to `on_before_batch_transfer` shouldn't be limited to whether any strategy/accelerator supports explicit transfer to batches to the device. Users should still be able to use this for batch transforms, etc...

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- x ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @awaelchli @ninginthecloud @rohitgr7 @akihironitta

### CODE REVIEW

File: tests/tests_pytorch/core/test_datamodules.py [hunk 1]
```diff
@@ -265,75 +265,6 @@ def test_full_loop(tmpdir):
     assert result[0]["test_acc"] > 0.6
 
 
-@pytest.mark.parametrize(
```
Comment (rohitgr7) [inline tests/tests_pytorch/core/test_datamodules.py:268]: removed this since this is almost the same as the one we have in `models/test_hooks.py::test_apply_batch_transfer_handler`

and we have unit tests for DataHookSelector which checks from which component the hooks is taken from.

File: src/pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -264,7 +264,9 @@ def advance(self) -> None:  # type: ignore[override]
         dataloader = self.trainer.train_dataloader
         assert self._data_fetcher is not None
         self._data_fetcher.setup(
-            dataloader, batch_to_device=partial(self.trainer._call_strategy_hook, "batch_to_device", dataloader_idx=0)
+            dataloader,
+            pre_batch_to_device=partial(self.trainer.lightning_module._on_before_batch_transfer, dataloader_idx=0),
+            batch_to_device=partial(self.trainer._call_strategy_hook, "batch_to_device", dataloader_idx=0),
```
Comment (carmocca) [inline src/pytorch_lightning/loops/fit_loop.py:269]: I don't think this should add a new argument, insted the `batch_to_device` here should be a local function that wraps both calls.

Comment (rohitgr7) [reply src/pytorch_lightning/loops/fit_loop.py]: done!

File: docs/source-pytorch/accelerators/ipu_basic.rst [hunk 3]
```diff
@@ -67,6 +67,5 @@ Please see the `MNIST example <https://github.com/Lightning-AI/lightning/blob/ma
 * Since the step functions are traced, branching logic or any form of primitive values are traced into constants. Be mindful as this could lead to errors in your custom code.
 * Clipping gradients is not supported.
 * It is not possible to use :class:`torch.utils.data.BatchSampler` in your dataloaders if you are using multiple IPUs.
-* IPUs handles the data transfer to the device on the host, hence the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_before_batch_transfer`,
-  :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device` and :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_after_batch_transfer`
-  do not apply here and if you have overridden any of them, an exception will be raised.
+* IPUs handles the data transfer to the device on the host, hence the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device` and
```
Comment (awaelchli) [inline docs/source-pytorch/accelerators/ipu_basic.rst:70]:
```suggestion
* IPUs handle the data transfer to the device on the host, hence the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device` and
```

**Review by awaelchli — APPROVED** [review]
Accept to unblock, but I disagree with it the same way I disagreed when the on_before and on_after hooks were originally introduced. 

`on_before_batch_transfer` has no meaning when the transfer is handled internally by the accelerator.

**Comment by rohitgr7** [comment]
> Accept to unblock, but I disagree with it the same way I disagreed when the on_before and on_after hooks were originally introduced.

> on_before_batch_transfer has no meaning when the transfer is handled internally by the accelerator.


we can probably remove them if you have better recommendations here on how users can perform batch transformations on/off device?

File: src/pytorch_lightning/core/module.py [hunk 4]
```diff
@@ -286,28 +286,29 @@ def loggers(self) -> List[Logger]:
         """Reference to the list of loggers in the Trainer."""
         return self.trainer.loggers if self._trainer else []
 
+    def _call_batch_hook(self, hook_name, *args):
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:289]: typing

File: src/pytorch_lightning/core/module.py [hunk 5]
```diff
@@ -286,28 +286,29 @@ def loggers(self) -> List[Logger]:
...
+                self._trainer._call_lightning_module_hook
+                if isinstance(obj, self.__class__)
+                else self._trainer._call_lightning_datamodule_hook
+            )
+            return trainer_method(hook_name, *args)
+        else:
+            hook = getattr(self, hook_name)
+            return hook(*args)
+
+    def _on_before_batch_transfer(self, batch: Any, dataloader_idx: int = 0):
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:303]: typing

File: src/pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 6]
```diff
@@ -143,10 +143,13 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
...
-            dataloader,
-            batch_to_device=partial(self.trainer._call_strategy_hook, "batch_to_device", dataloader_idx=dataloader_idx),
-        )
+
+        def batch_to_device(batch: Any, dataloader_idx: int) -> Any:
+            batch = self.trainer.lightning_module._on_before_batch_transfer(batch, dataloader_idx=dataloader_idx)
+            batch = self.trainer._call_strategy_hook("batch_to_device", batch, dataloader_idx=dataloader_idx)
+            return batch
+
+        self._data_fetcher.setup(dataloader, batch_to_device=partial(batch_to_device, dataloader_idx=dataloader_idx))
```
Comment (carmocca) [inline src/pytorch_lightning/loops/dataloader/evaluation_loop.py:152]: I don't think the `partial` is necessary anymore as it can be put inside the local function you just added above

```suggestion

        def batch_to_device(batch: Any) -> Any:
            batch = self.trainer.lightning_module._on_before_batch_transfer(batch, dataloader_idx=dataloader_idx)
            batch = self.trainer._call_strategy_hook("batch_to_device", batch, dataloader_idx=dataloader_idx)
            return batch

        self._data_fetcher.setup(dataloader, batch_to_device=batch_to_device
```

File: src/pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 7]
```diff
@@ -143,10 +143,13 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
         dataloader_idx = self.current_dataloader_idx
         dataloader = self.current_dataloader
         assert self._data_fetcher is not None
```
Comment (carmocca) [inline src/pytorch_lightning/loops/dataloader/evaluation_loop.py:145]: nit: can you keep this assert beside where we use it? Above L152 now

File: src/pytorch_lightning/loops/fit_loop.py [hunk 8]
```diff
@@ -263,9 +263,13 @@ def advance(self) -> None:  # type: ignore[override]
...
-        self._data_fetcher.setup(
-            dataloader, batch_to_device=partial(self.trainer._call_strategy_hook, "batch_to_device", dataloader_idx=0)
-        )
+
+        def batch_to_device(batch: Any, dataloader_idx: int) -> Any:
+            batch = self.trainer.lightning_module._on_before_batch_transfer(batch, dataloader_idx=dataloader_idx)
+            batch = self.trainer._call_strategy_hook("batch_to_device", batch, dataloader_idx=dataloader_idx)
+            return batch
+
+        self._data_fetcher.setup(dataloader, batch_to_device=partial(batch_to_device, dataloader_idx=0))
```
Comment (carmocca) [inline src/pytorch_lightning/loops/fit_loop.py:272]: Same comment about partial

**Review by carmocca — APPROVED** [review]
